pkg/resources/function_commons.go

```diff
@@ -0,0 +1,344 @@
+	}
+	sqlFunctionSchemaDefinition = functionSchemaDef{
+		additionalArguments:           []string{},
+		functionDefinitionDescription: functionDefinitionTemplate("SQL", "https://docs.snowflake.com/en/developer-guide/udf/sql/udf-sql-introduction"),
+	}
+)
+
+var (
+	javaFunctionSchema       map[string]*schema.Schema
+	javascriptFunctionSchema map[string]*schema.Schema
+	pythonFunctionSchema     map[string]*schema.Schema
+	scalaFunctionSchema      map[string]*schema.Schema
+	sqlFunctionSchema        map[string]*schema.Schema
+)
+
+// TODO [SNOW-1348103]: add null/not null
+// TODO [SNOW-1348103]: currently database and schema are ForceNew but based on the docs it is possible to rename with moving to different db/schema
+// TODO [SNOW-1348103]: copyGrants and orReplace logic omitted for now, will be added to the limitations docs
+// TODO [SNOW-1348103]: temporary is not supported because it creates a per-session object; add to limitations/design decisions
+var functionBaseSchema = map[string]schema.Schema{
+	"database": {
+		Type:             schema.TypeString,
+		Required:         true,
+		ForceNew:         true,
+		DiffSuppressFunc: suppressIdentifierQuoting,
+		Description:      blocklistedCharactersFieldDescription("The database in which to create the function."),
+	},
+	"schema": {
+		Type:             schema.TypeString,
+		Required:         true,
+		ForceNew:         true,
+		DiffSuppressFunc: suppressIdentifierQuoting,
+		Description:      blocklistedCharactersFieldDescription("The schema in which to create the function."),
+	},
+	"name": {
+		Type:             schema.TypeString,
+		Required:         true,
+		Description:      blocklistedCharactersFieldDescription("The name of the function; the identifier does not need to be unique for the schema in which the function is created because UDFs are identified and resolved by the combination of the name and argument types. Check the [docs](https://docs.snowflake.com/en/sql-reference/sql/create-function#all-languages)."),
+		DiffSuppressFunc: suppressIdentifierQuoting,
+	},
+	"is_secure": {
+		Type:             schema.TypeString,
+		Optional:         true,
+		Default:          BooleanDefault,
+		ValidateDiagFunc: validateBooleanString,
+		DiffSuppressFunc: IgnoreChangeToCurrentSnowflakeValueInShow("is_secure"),
+		Description:      booleanStringFieldDescription("Specifies that the function is secure. By design, the Snowflake's `SHOW FUNCTIONS` command does not provide information about secure views (consult [function docs](https://docs.snowflake.com/en/sql-reference/sql/create-function#id1) and [Protecting Sensitive Information with Secure UDFs and Stored Procedures](https://docs.snowflake.com/en/developer-guide/secure-udf-procedure)) which is essential to manage/import function with Terraform. Use the role owning the function while managing secure functions."),
```

Comment:
Secure views?

Reply:
Fixed in https://github.com/Snowflake-Labs/terraform-provider-snowflake/pull/3263